torch_np/_wrapper.py

```diff
@@ -310,17 +313,22 @@ def empty_like(prototype, dtype=None, order="K", subok=False, shape=None):
     return result
 
 
+@_decorators.dtype_to_torch
 def full(shape, fill_value, dtype=None, order="C", *, like=None):
     _util.subok_not_ok(like)
     if order != "C":
         raise NotImplementedError
-    if isinstance(fill_value, ndarray):
-        fill_value = fill_value.get()
+
+    fill_value = asarray(fill_value).get()
     if dtype is None:
-        torch_dtype = asarray(fill_value).get().dtype
-    else:
-        torch_dtype = _dtypes.torch_dtype_from(dtype)
-    return asarray(torch.full(shape, fill_value, dtype=torch_dtype))
+        dtype = fill_value.dtype
```

Comment:
A better way to implement this would be having a function that given a tensor or a number returns its torch dtype. The reason here is that we don't want to wrap scalars into tensors if we don't need to. Scalars are easier to treat in a compiler than tensors, and often get extra optimisations (no memory allocations, no synchronisations when accessed if we are working on CUDA, you can specialise on them...)

Note that this applies in general, so we should be mindful about this going all across.

Reply:
Let's move this to https://github.com/Quansight-Labs/numpy_pytorch_interop/issues/56, to not conflate a small cleanup of this PR with a rather large rework.